feat: [HUDI-9553] Add equals and hashCode override to Predicate inner classes

voonhous opened the pull request:
### Describe the issue this Pull Request addresses

Closes: #17072

The `Predicate` inner classes in `Predicates` did not override `equals`/`hashCode`, so equivalent predicates compared by identity and were never equal (e.g. in unit tests).

### Summary and Changelog

Add value-based `equals`/`hashCode` to the `Predicate` inner classes. Per review feedback, these are generated via Lombok's `@EqualsAndHashCode` instead of hand-written boilerplate:

- `BinaryExpression` is annotated with `@EqualsAndHashCode`; its subclasses (`And`, `Or`, `StringStartsWith`, `StringContains`, `BinaryComparison`) use `@EqualsAndHashCode(callSuper = true)` to compare the shared `left`/`operator`/`right`.
- `In`, `IsNull`, `IsNotNull`, `Not`, `StringStartsWithAny` use `@EqualsAndHashCode`.
- `TrueExpression`/`FalseExpression` keep manual overrides: as field-less singletons, Lombok would emit a constant `hashCode` and collide the two classes.

### Impact

No functional impact. Users can now compare `Predicate` inner classes with `equals` (e.g. in unit tests).

### Risk Level

none

### Documentation Update

none

### Contributor's checklist

- [x] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [x] Enough context is provided in the sections above
- [x] Adequate tests were added if applicable

voonhous committed:
[HUDI-9553] Add equals and hashCode override to Predicate inner classes

yihua reviewed (commented):
@voonhous should we use lombok for auto codegen of `equals` and `hashCode` given that lombok is included?

yihua commented on hudi-common/src/main/java/org/apache/hudi/expression/Predicates.java:
```diff
@@ -120,6 +121,19 @@ public <T> T accept(ExpressionVisitor<T> exprVisitor) {
     }
+
+    @Override
+    public boolean equals(Object o) {
```
Just to flag — since `Literal` doesn't override `equals`/`hashCode`, these new predicate `equals` methods still rely on reference equality for child expressions. So `assertEquals` will pass in your tests because you're reusing the same `Literal` instances, but two separately constructed `Predicates.eq(new Literal<>("a", StringType), new Literal<>(10, IntType))` would still fail `equals` even though they're structurally identical. You may want to add `@EqualsAndHashCode` to `Literal` (and possibly `NameReference`) to get the full benefit.

voonhous committed:
[HUDI-9553] Use Lombok @EqualsAndHashCode for Predicate inner classes

Replace the hand-written equals/hashCode in Predicates with Lombok
@EqualsAndHashCode, per review feedback on #13481.

- Add @EqualsAndHashCode to BinaryExpression so its subclasses (And, Or,
  StringStartsWith, StringContains, BinaryComparison) get equality over the
  shared left/operator/right fields via callSuper = true.
- Annotate the self-contained predicates (In, IsNull, IsNotNull, Not,
  StringStartsWithAny) with @EqualsAndHashCode.
- Keep TrueExpression/FalseExpression manual: as field-less singletons,
  Lombok would emit a constant hashCode and collide the two classes.
- Fix import ordering in PredicatesTest.

voonhous commented:
@wombatu-kun Can i get your review on this too? Thank you.

wombatu-kun commented on hudi-common/src/test/java/org/apache/hudi/expression/PredicatesTest.java:
```diff
@@ -0,0 +1,239 @@
+class PredicatesTest {
+
+  // Define some common literal expressions for use in tests
+  private final Expression a = new Literal<>("a", Types.StringType.get());
```
Every assertEquals here reuses the same operand instances (a, b, c), so operand equality is never exercised by value. Literal and NameReference do not override equals (Literal.java:35, NameReference.java:28 carry only @AllArgsConstructor/@Getter), so the predicate @EqualsAndHashCode compares operands by identity: two predicates built from distinct-but-equal operands are unequal - exactly the independently-built-tree case the Trino Expression Index comparison relies on. This test fails today, and passes once @EqualsAndHashCode is added to the leaf types Literal and NameReference:

```java
@Test
void testEqualsWithDistinctButEqualOperands() {
  Expression a1 = new Literal<>("a", Types.StringType.get());
  Expression a2 = new Literal<>("a", Types.StringType.get());
  assertEquals(isNull(a1), isNull(a2));
}
```

voonhous replied:
Good catch. You're right that with the leaf types comparing by identity, independently-built trees never match, which defeats the purpose here.

Added `@EqualsAndHashCode` to `Literal` and `NameReference`, plus a `testEqualsWithDistinctButEqualOperands` case that builds distinct-but-equal operands (your `isNull` example + a nested `and(eq(...), isNotNull(...))` tree over both leaf types).

voonhous committed:
[HUDI-9553] Add @EqualsAndHashCode to Literal and NameReference leaf operands

Predicate @EqualsAndHashCode compared operands by identity, because the leaf
expression types Literal and NameReference did not override equals/hashCode.
Independently built predicate trees with value-equal operands - the case the
Trino Expression Index comparison relies on - were therefore unequal.

Add @EqualsAndHashCode to both leaf types, plus a PredicatesTest case that
builds distinct-but-equal operands so equality is exercised by value rather
than by identity.
